### PR TITLE
MVC Foundations

### PR DESCRIPTION
**Pull request by wfrece**

### CODE REVIEW

File: model/spectra/question.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+# Defines a request for input from respondents (ie.e, a form field definition)
+class Question(Entity):
+    def __init__(self, **kwargs):
+        self.value_type = value_type # how we should interpret the answer
+        self.validation_criteria = validation_criteria # how we determine if we accept the answer (null for type default)
+        self.answer_key = answer_key # how we should evaluate the answer
+        # template?
+
+from marshmallow import Schema, fields
```
Comment (dkheyman) [inline model/spectra/question.py]: Is there a reason this is here and not at the top? A bit hard to follow what this file really is =]

File: model/spectra/subject.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+from ../common import Entity
+
+# Represents an entity about which an answer can be given
+class Subject(Entity):
+    def __init__(self, **kwargs):
```
Comment (dkheyman) [inline model/spectra/subject.py]: I am guessing this is just a placeholder and will be continued?

File: model/spectra/question.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+class UserSchema(Schema):
```
Comment (dkheyman) [inline model/spectra/question.py]: Can you explain what the difference is between an Entity and a Schema? and why are these in the same file?

File: view/__init__.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from flask_restplus import Api
+from view.analytics import analytics_ns
+from view.distribution import distribution_ns
+
+api = Api(version='1.0', title='Spectra API',
+    description='A RESTful API for the Spectra platform',
+    doc='/spectra/api'
+)
+api.add_namespace(analytics_ns)
+api.add_namespace(distribution_ns)
```
Comment (dkheyman) [inline view/__init__.py]: So as we add new namespaces, we just establish them as a view, and add them here?

File: controller/distribution.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from controller.plugins.distributor import Distributor
+
+class DistributionController():
+    def __init__(self, **kwargs):
+        for key, value in kwargs.iteritems():
+            setattr(self, key, value)
+
+    def fetch_distribution(self, student_ids, student_count, assignments_per_student, initial_seed):
```
Comment (dkheyman) [inline controller/distribution.py]: I would like to make amends to this branch to fix my naming of things and the namespace I created. Can I do that before you merge into master, you think?

File: controller/analytics.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+from controller.common.db_tools import DBManager
+from controller.plugins.mastery import Mastery
+from controller.plugins.judgment import Judgment
+
+class AnalyticsController():
+    def __init__(self, **kwargs):
+        for key, value in kwargs.iteritems():
+            setattr(self, key, value)
+
+        self.dbm = DBManager(db_host = os.environ.get('SPECTRA_DB_HOST', 'localhost'),
```
Comment (dkheyman) [inline controller/analytics.py]: Should this DBManager object live at a class above the Analytics controller? Because the Distribution Controller will probably also have to use it, to make queries about respnodents and subjects. Thoughts?

File: view/analytics.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+from flask import request, render_template, url_for, jsonify
+from flask_restplus import Namespace, Resource, fields
+from controller.analytics import AnalyticsController
+import pprint
+pp = pprint.PrettyPrinter(indent=4)
+
+a = AnalyticsController()
+analytics_ns = Namespace('spectra/api/v1/analytics', description='Spectra Analytics Engine (SAE) Operations')
```
Comment (dkheyman) [inline view/analytics.py]: I like this naming a lot!

File: view/__init__.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+from flask_restplus import Api
+from view.analytics import analytics_ns
+from view.distribution import distribution_ns
+
+api = Api(version='1.0', title='Spectra API',
+    description='A RESTful API for the Spectra platform',
+    doc='/spectra/api'
```
Comment (dkheyman) [inline view/__init__.py]: What is this doc argument? Is it a special namespace of some sort?

File: view/analytics.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+# NOTE: Trailing slash prevents unnecessary 404s, but it implies 3** redirects)
+# NOTE: Possible converter types for embedded URI params: string(default)|int|uuid
+@analytics_ns.route('/mastery')
+class MasteryEndpoint(Resource):
+    def get(self):
+        question_id     = request.args.get('question_id',       '')
+        questionnaire_id= request.args.get('questionnaire_id',  '')
+        instructor_id   = request.args.get('instructor_id',     '')  # instructor respondent ID
```
Comment (dkheyman) [inline view/analytics.py]: Hmm, maybe it's time we reconsider calling this the instructor? Maybe we can say the 'truth_responderId' to get away from the classroom example?

File: view/analytics.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            debug               = debug
+        )
+        return jsonify(results)
+
+@analytics_ns.route('/judgment')
+class JudgmentEndpoint(Resource):
+    def get(self):
+        question_id      = request.args.get('question_id',       '')
+        questionnaire_id = request.args.get('questionnaire_id',  '')
+        instructor_id    = request.args.get('instructor_id',     '') # instructor respondent ID
```
Comment (dkheyman) [inline view/analytics.py]: Same point as above.

File: controller/plugins/mastery.py [hunk 11]
```diff
@@ -5,6 +5,18 @@
 import pprint
 pp = pprint.PrettyPrinter(indent=4)
 
+'''
+mastery_req_parser = api.parser()
```
Comment (dkheyman) [inline controller/plugins/mastery.py]: Why is this commented out?

File: model/spectra/question.py (hunk 1 above)
Comment (wfrece) [reply model/spectra/question.py]: Remember that i mentioned the PR was 80% done. The part that hasn't been applied yet is the marshmallow schemas. I've actually fixed this in my latest code...it was just here as a placeholder. plz ignore its placement for now.

File: model/spectra/subject.py (hunk 2 above)
Comment (wfrece) [reply model/spectra/subject.py]: No, actually. This is all there is to a "subject" object. Same as respondent. Remember? It need only "exist" - all meaning is from mappings from Fairgrade tier.

File: model/spectra/question.py (hunk 3 above)
Comment (wfrece) [reply model/spectra/question.py]: Entity is a base object from which all other spectra model objects are drived. The Schema is unrelated...its a marshmallow concept that applies serialization types to native python data types and provides validation methodologies and quick type-checking when dealing with inputs and outputs (which is why, coincidentally, the [parsing bit of swagger - see top of page](https://flask-restplus.readthedocs.io/en/stable/parsing.html) is being removed in place of marshmallow.

File: view/__init__.py (hunk 4 above)
Comment (wfrece) [reply view/__init__.py]: yes, new families of endpoints. you declare the namespace in the view itself, and then "register" the view here to link it into the swagger, which can be found at /spectra/api
